Attempt to implement a "modern" fortran interface

GiovanniBussi opened the pull request:
This is an attempt to write a "modern Fortran" interface (wrt to the current "classic Fortran" interface). In particular, there is now a dedicated `type(plumed)` as a handler to the plumed instance. In the classic interface, this has been stored in a `character(32)` variable. Advantages would be:

1. Clearer code and type checking. In particular, passing a `type(plumed)` to a subroutine would have an obvious meaning. This is in common with C, C++, and Python wrappers, where the plumed instance is a struct (C) or a class (C++/Python).
2. Type-bound methods: `p%cmd("setPositions",positions)` instead of `plumed_f_cmd(p,"setPositions",positions)`. This is mostly aesthetic, but has some advantage in that is it more compact. This is in common with C++ and Python wrappers. C wrappers instead use plain functions (`plumed_cmd`).
3. Constructors/destructors/assignment. Ideally, we would like to have the plumed instance to be automatically initialized and finalized when the object goes out of scope. A new reference should be added when using assignment operator. The reference counter is already embedded in PLUMED. This is in common with C++ wrappers. In Python wrappers we don't do it since the reference counting is done by Python. In C wrappers the user has to explicitly create and finalize plumed structures, and the reference counter can be increased with `plumed_create_reference`. In practice, in C the rule is that for every `plumed_create*` call there should be a `plumed_finalize` call on the returned object. Management is explicitly done by the user.

Points 1 and 2 easy to obtain. Point 3 is more tricky.

In this pull request I made a tentative implementation. On top of this I could easily add exception handling (as [here](https://github.com/plumed/plumed2/pull/698#issuecomment-859659144)). I am still struggling a bit with point 3 above, so I ask once more @aradi's advise on this.

**FORTRAN does not have constructors**. One recommendation that I've seen around is to use a function returning a copy of the object. I didn't use this idea for two reasons: (a) the function should be called explicitly, so it has no advantage on the classic implementation, and (b) I have some problems with returning plumed objects from functions correctly (see below). Another possible solution is to have a `logical :: initialized` variable in the plumed type that tells if the object has been initialized. At first usage, if it has not been initialized, the constructor is called (and in turn calls `plumed_f_create`). This is what I used here (see e.g. [these lines](https://github.com/plumed/plumed2/blob/7579158b3f7fe4ecc533d7544a0989e2d9dc7c93/fortran/plumed.f90.template#L330-L332)).

**FORTRAN do have destructors (called "finalizers")**. The destructor can check if the object was initialized. [If yes](https://github.com/plumed/plumed2/blob/7579158b3f7fe4ecc533d7544a0989e2d9dc7c93/fortran/plumed.f90.template#L281-L295), then it calls `plumed_f_finalize`.

**FORTRAN assignment can be overloaded**. This is crucial to avoid users taking a copy of an object that is then automatically finalized. I made the `=` operator call the function `plumed_f_create_reference`, that [creates a new reference and increases the internal reference counter](https://github.com/plumed/plumed2/blob/7579158b3f7fe4ecc533d7544a0989e2d9dc7c93/fortran/plumed.f90.template#L319-L324). This is exactly what happens now in the [C++ wrappers](https://github.com/plumed/plumed2/blob/master/src/wrapper/Plumed.h#L1862-L1868). Notice that in C++ one should also explicitly destroy the calling object (i.e., `A`, in `A=B`) whereas I've seen empirically than Fortran does it without the need to write it.

Consider this code:

```
type(plumed) :: p
call p%create()
call p%cmd("init")
call p%finalize()
```
The call to `create` is superfluous: if one forgets it, it is called automatically by `cmd`. Also the call to `finalize` is superfluous: if one forgets it, is it called by the destructor of the class. Thus, the following version works as well

```
type(plumed) :: p
call p%cmd("init")
```
So far so good. However, there are corner cases where my implementation does not work. In particular, the tricky thing is to make sure the destructor is called. This is important since it is required to finalize the PLUMED object, free memory, and write some information in the log file.

I identified the following cases where the destructor would not be called, but there might be others.

**A. Using a plumed object from PROGRAM**

In the following code:

```
program main
use plumed_module
implicit none
type(plumed) :: p
call p%cmd("init")
end program main
```
The destructor is not called. This seems a feature rather than a bug: variables in main are supposed to never go out of scope in Fortran. Coming from C++ this is difficult to understand, but that's it. Maybe this is well known to Fortran people and acceptable.

**B. Returning a plumed object from a function**

In the following code:

```
function f() result p
type(plumed) :: p
call p%create()
end function f

function g()
type(plumed) :: p
p=f()
end function g
```
With gfortran the destructor is not called on the result of `f()`. This seems a known bug of gfortran, but this thing is not completely clear to me. I can confirm that ifort works as I would expect here (thus destroying the temporary variable). No idea about other compilers.

**C. Other?**

While working on this, I discovered that the destructor should be `impure elemental`, otherwise it would not be called if one allocates an array of `type(plumed)` instances. This is [fixed now](https://github.com/plumed/plumed2/blob/7579158b3f7fe4ecc533d7544a0989e2d9dc7c93/fortran/plumed.f90.template#L290-L295), but I don't know if there are other situations where destructors or constructors are not called 

**What's next**

If I cannot solve the problems above, I am a bit skeptical in having automatic construction and destruction implemented if there are cases where they do not work. I would rather force the user to explicitly call `create` and `finalize`, and would let `=` copy the handle without increasing the reference counter. This would be identical to the current C interface (and different from the C++ and Python interfaces): a dedicated struct/type to hold the PLUMED instance that needs to be explicitly created and destroyed. In this case I am not sure if the added value of type-bound methods is sufficient.

##### Target release

<!-- please tell us where you would like your code to appear (e.g. v2.4): -->
I would like my code to appear in release __v2.8__

GiovanniBussi commented:
Note: while running on GitHub Actions, I see that `final` statement is not supported by gfortran 4 (Centos7 build).

danielhollas commented:
@GiovanniBussi thank you for this effort and thanks for tagging me on the previous PR. I don't have anything clever to say at this point. imo the stronger type checking is the biggest win, and the other stuff is more nice-to-have, as you point out.

It would be great if this new interface worked with GFortran 7, since that is the version we have on our clusters (and updating it might be nontrivial).

aradi commented:
First: Is copying an instance, which is worth to bother with? Is it really realistic, that you want to have several copies of the same plumed instance? Even if you override the assignment operator, there are several ways to circumvent the reference counting, e.g. by sourced allocation `allocate(newinst, source=oldinst)`, which would create a bitwise copy without calling the user-defined assignment, killing your ref

Second: I would not use a constructor function, as it makes extra unnecessary finalizations and copies. Some projects use special named subroutines to do the task, such as `call new_plumed(...)` or `call plumed_init()`, I find those solutions much clearer, simpler and more robust than a type constructor functions.

Third: If you intend to support gfortran 4 (which you should definitely NOT), then forget any OO mambo-jumbo. I'd suggest to make the F03 interface optional at build (with the possibility of disabling its build), and go for at least gfortran 9 and above. From there on, you have a decent compiler support (with some dark edges of finalization, which may be still broken in gfortran). GFortran 7 may or may not work, but I am sorry to say, but an HPC cluster with only 4 years old compilers should look for a better maintainer. And since you offer Plumed via Conda, everyone can get binary packages without the hassle of compilation. (And Conda just changed to GFortran 9 globally.)

Fourth: I personally would favor not to have to call the destructor explicitly. This is something you typically would forget anyway, leading to potential memory holes.

Later today, I'll try to have a look on the source code and comment on that in detail.

aradi commented:
Just to extend: I was really not intended to be harsh about clusters with old build environments. Still, I do think it is reasonable to assume, that  on a well maintained cluster every 2nd year somebody takes the effort to set up a recent compiler and a derived toolchain. And if the new interface is optional, people still can build Plumed without it. And everybody not an expert should use anyway conda to get the package IMO. (This is also the recommended way of getting DFTB+ nowadays...). It would be a pitty, to design an "modern" interface, which just works like the C-interface (which means basically like the ancient Fortran interface with a little bit of syntactic sugar), just because you want to support very old deprecated compilers. The oldest compiler being actively maintained is gcc 9, everything older does not even get bug.fixes any more. There were so many improvements and bug-fixes in the last few years, that it would be a pity not to profit from them.

GiovanniBussi commented:
Thanks a lot! A few quick comments below before you review the code.

> First: Is copying an instance, which is worth to bother with? Is it really realistic, that you want to have several copies of the same plumed instance? Even if you override the assignment operator, there are several ways to circumvent the reference counting, e.g. by sourced allocation `allocate(newinst, source=oldinst)`, which would create a bitwise copy without calling the user-defined assignment, killing your reference counting anyway.

What I want is not a way to copy an instance but to copy a reference. If I don't do this, there's no way to implement a reasonable destructor (see last point of this message). I mean, anyone taking a raw copy (with operator`=`) of a PLUMED instance would then finalize it twice, leading to memory corruption. Maybe I am too used to C++ and not to Fortran. Thinking in C++ terms, I would have two options: (a) make it un-copyable, so that people do not make mistake (as a `std::unique_ptr`) or (b) make it copyable, using the reference counter (as a `std::shared_ptr`). If there's a way to *disable* assignment operator in Fortran, that would be equally safe I think. In the worst case, I can do it at runtime, redefining assignment operator as `stop`.

> 
> Second: I would not use a constructor function, as it makes extra unnecessary finalizations and copies. Some projects use special named subroutines to do the task, such as `call new_plumed(...)` or `call plumed_init()`, I find those solutions much clearer, simpler and more robust than a type constructor functions.

I agree, this is what I did. And I also made this initialization optional, by calling it automatically on first usage.

> 
> Third: If you intend to support gfortran 4 (which you should definitely NOT), then forget any OO mambo-jumbo. I'd suggest to make the F03 interface optional at build (with the possibility of disabling its build), and go for at least gfortran 9 and above. From there on, you have a decent compiler support (with some dark edges of finalization, which may be still broken in gfortran). GFortran 7 may or may not work, but I am sorry to say, but an HPC cluster with only 4 years old compilers should look for a better maintainer. And since you offer Plumed via Conda, everyone can get binary packages without the hassle of compilation. (And Conda just changed to GFortran 9 globally.)

Yes it's fine to disable this test on Centos7. I totally agree on using modern Fortran for this interface. Even committing to Fortran 2018 would be ok, or at least to the subset of Fortran 2018 supported by a given gfortran version.

> 
> Fourth: I personally would favor not to have to call the destructor explicitly. This is something you typically would forget anyway, leading to potential memory holes.'

Totally agree, and that's why I was trying to implement this.

> 
> Later today, I'll try to have a look on the source code and comment on that in detail.

Looking forward for this! Thanks again

GiovanniBussi commented:
Some further sources:

- The fact that objects declared in main are not finalized is discussed [here](https://stackoverflow.com/questions/25907363/deallocating-memory-with-destructor-in-fortran). If I trust this source, this (I think odd) behavior is written in the Fortran standard and thus not expected to change.

- The fact that the gfortran behavior on objects returned from functions is a bug with respect to the standard is reported for instance [here](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=37336#c30) (found from [this](https://stackoverflow.com/questions/59985499/are-fortrans-final-subroutines-reliable-enough-for-practical-use)). The open case is `b) Finalization of functions results after their use`. Point `a) Finalization of the LHS during intrinsic assignment` seems fixed. I don't understand if `c) Finalization of structure/array constructors after their use.` is relevant in this case.

And I should add that calling the destructor in our case is not just a matter of releasing memory, but also of letting know PLUMED that the calculation is over and write on the log file the timing, plus possibly releasing other resources.

aradi reviewed (commented):
I made a few suggestions. Not in the template, but in the generated code as it is easier to recognize the purpose, but you would have to apply them in the template, of course.

A few general remarks:
* I'd suggest to have two different modules for the two different interfaces, e.g. `plumed_module` for the legacy one and something like `plumed08_module` for the Fortran 2008 one. (It is done similarly with mpi). People should import the one or the other and not mix those interfaces. If they are both in the same module, the temptation is bigger.

* I am not a big fan of type bound constructors, due to their virtuality, as explained in the comments. However, there is no established standard way of handling this. The projects I have some influence on definitely avoid to use them :wink: 

* The user defined assignment is OK. However, keep in mind, that it is still super-easy to circumvent it with a sourced allocation (`allocate(newinst, source=oldinst)`). But his is something you can not do anything about in Fortran.

* As your question about finalization in programs: Yes, programs and modules do not finalize their "global" variables at exit. You can trigger finalization by making the variable `allocatable` and deallocate it explicitely before exit, but that's a pain. In Fortran, you are not expected to do put anything essential/functional into a destructor (like writing out additional information), but only cleaning up instructions, which would be automatically executed by the framework when you exit the program anyway (like closing files, freeing all memory allocations, etc.).

GiovanniBussi commented on fortran/plumed.f90:
```diff
@@ -23,6 +23,92 @@ module plumed_module
+  public :: plumed
+
+  type plumed
+    character(kind=c_char,len=32) :: handle
+    logical                       :: initialized = .false.
```
ok

GiovanniBussi commented on fortran/plumed.f90:
```diff
@@ -23,6 +23,92 @@ module plumed_module
+  public :: plumed
+
+  type plumed
+    character(kind=c_char,len=32) :: handle
```
I understand but at this level I prefer to avoid code duplication and base this interface on the other. I cannot do the other way round since not all compilers will support this interface.

GiovanniBussi commented on fortran/plumed.f90:
```diff
@@ -408,6 +494,374 @@ end subroutine plumed_f_gcmd_safe_long_double
 
+     subroutine pl_create(this)
+       class(plumed), intent(inout) :: this
+       call this%finalize()
```
Good point. I didn't know this, thanks!

GiovanniBussi commented:
OK I implemented all the suggestions thanks! There still a few points.

**Non-type-bound constructor**

I think your point is important since the main reason to have OOP is to allow others to inherit from this, so I would like to use the best syntax now. We need constructors to allow users to customize the way the new object is created, but a user could avoid them at all. Indeed, the default one (`p%create()`) is called at first use. In order to accomodate your suggestion I could just replace
```
call p%create_dlopen("/path/to/libplumedKernel.so")
```
with
```
call plumed_create_dlopen(p,"/path/to/libplumedKernel.so")
```
Do you think it would be clearer/more familiar for Fortran users?

**Non-destroyed global data**

I see your point. I added a warning in the PLUMED library so that, when the run finishes cleanly, it gives a warning on stderr if there is a dangling PlumedMain instance. This will at least serve to developers as a reminder that it is not good practice not to finalize objects (also calling PLUMED from languages other than Fortran).

**Non-destroyed objects when returned from functions**

This is a very tricky issue, since the Fortran standard requires this but the bug is present in gfortran 10. As a workaround, I added functions to manipulate by hand the reference counter (`incref` and `decref`, we have them in C++ already, they are there for troubleshooting), with as special option to decrease the reference counter until a specified number of references. The recommended workaround would then be to call
```
call p%decref(to=1)
```
on any object returned by a function (as [here](https://github.com/plumed/plumed2/blob/f043395aeca3853b5e7cd04b37f3d381104f6f89/regtest/basic/rt-make-fortran/testobj.f90#L48-L50)). This should work both with gfortran and with conforming compilers such as intel (I will wait for GitHub Actions before betting on this).

**Non-destroyed objects in XX situation**

If you are aware of any other case where objects are not finalized (either by the standard or by very common compilers being buggy) please let me know.

**Support for older gfortran**

(this is a TODO for me): fix the regression tests so that they can run with old compilers skipping the inclusion of the f08 module. We definitely plan to continue support for older gfortran with the classic interface @danielhollas, I need to make sure this is regularly tested on our CI.

aradi commented:
> 
> I think your point is important since the main reason to have OOP is to allow others to inherit from this, so I would like to use the best syntax now. We need constructors to allow users to customize the way the new object is created, but a user could avoid them at all. Indeed, the default one (`p%create()`) is called at first use. In order to accomodate your suggestion I could just replace
> 
> ```
> call p%create_dlopen("/path/to/libplumedKernel.so")
> ```
> 
> with
> 
> ```
> call plumed_create_dlopen(p,"/path/to/libplumedKernel.so")
> ```

Yes, that is IMO a much better way. Two remarks:

* I would use `init` instead of `create`, as the type is created (allocated) already, when you call this function, you merely initialize its content. (And would also fit as antonym to "final" or "finalization"...)

* I like this approach a lot. The signature would be then
  ```
  subroutine plumed_init_dlopen(this, dlpath)
    type(plumed), intent(out) :: this
    ...
  ```
  But to give the full picture, it comes with a disadvantage: people would have to write their own  initialization routines, when they extend the type, as it won't be inherited. Even, if they do not want to change the constructor signature at all. For example:
  ```
  type, extends(plumed) :: customized_plumed
  end type customized_plumed
  ...
  subroutine customized_plumed_init_dlopen(this, dlpath)
     type(customized_plumed), intent(out) :: this
     ...
     call plumed_init(this%plumed)
     ...
   end subroutine customized_plumed_init_dlopen
  ```
  But I don't know about any better way to circumvent it. Either, you fix the initialization signature for all derived instances making it basically impossible to change it or or every  derived instance must explicitey declare its own one(s).

> Do you think it would be clearer/more familiar for Fortran users?

Hard to say, there is no consensus on this. You can find both approaches in the Fortran world. [json-fortran](https://github.com/jacobwilliams/json-fortran/wiki/Example-Usage) uses your current approach, DFTB+ or [tblite](https://github.com/awvwgk/tblite/blob/main/src/tblite/classical/halogen.f90) use the approach described above. So, its basically a decision you have to make.

> **Non-destroyed global data**
> 
> I see your point. I added a warning in the PLUMED library so that, when the run finishes cleanly, it gives a warning on stderr if there is a dangling PlumedMain instance. This will at least serve to developers as a reminder that it is not good practice not to finalize objects (also calling PLUMED from languages other than Fortran).

Looks as a reasonable approach to me.

> **Non-destroyed objects when returned from functions**
> 
> This is a very tricky issue, since the Fortran standard requires this but the bug is present in gfortran 10. As a workaround, I added functions to manipulate by hand the reference counter (`incref` and `decref`, we have them in C++ already, they are there for troubleshooting), with as special option to decrease the reference counter until a specified number of references. The recommended workaround would then be to call

My advice considering the current status of the compilers is not to use functions which have return types with finalization. Is there any case, where you need a function returning a `type(plumed)` instance? You can/should always turn them into subroutines, which are must more robust (and you would spare unnecessary finalization calls anyway).

> **Non-destroyed objects in XX situation**
> 
> If you are aware of any other case where objects are not finalized (either by the standard or by very common compilers being buggy) please let me know.

The standard is waterproof as far as I can see and most compilers get most things right by now. But you always may hit some bugs ...

aradi commented:
One additional comment about initializers: You are providing 4, but are those all really needed? One could simplify the initialization by defining only one initializer:
```
subroutine plumed_init(this, path)
  type(plumed), intent(out) :: this
  character(kind=c_char,len=*), optional, intent(in)  :: path

  if (present(path)) then
     call plumed_f_create_dlopen(path // c_null_char,this%handle)
  else
    call plumed_f_create(this%handle)
  end if

end subroutine plumed_init
```
The `pl_create_reference` is fully equivalent to the user defined assignment, so no need to repeat it as initialiser IMO. And finally, I currently dont' see possible use cases of initializing an invalid plumed instance, but you may teach me about this.

Having one initialiser only, extending types would have only to provide one trivial wrapper, in case they don't want to change the API of the initialiser, as they could profit from Fortrans capability of forwarding optional arguments directly:
```
 subroutine customized_plumed_init(this, path)
  type(plumed), intent(out) :: this
  character(kind=c_char,len=*), optional, intent(in)  :: path

   call plumed_init(this%plumed, path)
```
This would be probably a fair price for allowing extending classes to change the API in case they need.

GiovanniBussi commented:
>I would use init instead of create, as the type is created (allocated) already, when you call this function, you merely initialize its content. (And would also fit as antonym to "final" or "finalization"...)

I see your point. However, we already use "init" as one of the possible key arguments to `cmd`. In addition, I would prefer to keep the names that we have in the C interface to avoid confusion.

>One additional comment about initializers: You are providing 4, but are those all really needed? One could simplify the initialization by defining only one initializer:

You are right. Indeed I was going to remove `create_reference` for the reason you said (after having verified that assignment works, and after having added `decref` to fix problematic situations). `create_invalid` is also only useful in C++, where construction happens immediately at declaration, in case you want to avoid a real construction (with the subsequent library loading) until a specified point in your program. E.g., you have an invalid object created at the beginning and replace it with a valid one once you decide to enable plumed. But it's not needed in Fortran, where construction is anyway delayed.

What you suggest would be identical to what we have in python, except that the additional argument is called `kernel` ([here](https://github.com/plumed/plumed2/blob/cc47166331617ae5f967b5727033884468f5ecbb/python/plumed.pyx#L52)), so would call it the same for consistency.

GiovanniBussi commented:
@aradi I implemented the pattern you recommended here: 0b13ebf74397bb74437f52b090113ba361c9a06f

For symmetry, I also made `plumed_finalize` unbound. It remains true that both `plumed_create` and `plumed_finalize` might be unnecessary for most usages: the former is called at first use (unless one wants to specify a path for dlopen) and the latter is only needed for special cases such as as finalizing a global instance (in main or in a module).

Could you please double check if it makes sense to have most of these functions as `impure elemental`?

If this is ok, I think I am ready to merge. 

I'll be back with another PR on dealing with exceptions, though on that side the work was more or less done.

Thanks again for your precious suggestions!

aradi commented:
> For symmetry, I also made `plumed_finalize` unbound. It remains true that both `plumed_create` and `plumed_finalize` might be unnecessary for most usages: the former is called at first use (unless one wants to specify a path for dlopen) and the latter is only needed for special cases such as as finalizing a global instance (in main or in a module).

While it becomes symmetric when you make `plumed_finalize` unbound, it puts extra burden on those, who wish to extend your class. If they wanted to provide an explicit finalizer, they would have to implement their own `extended_plumed_finalize()` routine. If you make it type bound, it could be automatically inherited. I think, one can safely assume, that all derived classes would wish to have the same API for the explicit finalizer, so that the constraints posed by the virtuality of the type bound procedure won't affect negatively the type extensions in this case (in contrast to the initializer routine).

> Could you please double check if it makes sense to have most of these functions as `impure elemental`?

Yes, it is, otherwise you would have to implement the destructors for all possible ranks, or hope, that the users would take care of it manually.

> I'll be back with another PR on dealing with exceptions, though on that side the work was more or less done.

Yes, we definitely need a proper error reporting mechanism, so that the caller can exit gracefully.

> Thanks again for your precious suggestions!

You're very welcome. I am always very happy to see, if a project starts to use all the nice feature current Fortran offers, instead of getting stuck on the level of a 44 years old standard...
